pandas/core/reshape.py

```diff
@@ -926,38 +929,139 @@ def wide_to_long(df, stubnames, i, j):
+       A(quarterly)2010  A(quarterly)2011  B(quarterly)2010
+    0            0.531828            0.724455            0.322959
+    1            0.634401            0.611024            0.361789
+    2            0.849432            0.722443            0.228263
+
+       B(quarterly)2011  X  id
+    0            0.293714  0   0
+    1            0.630976  2   1
+    2            0.092105  2   2
+
+    >>> pd.wide_to_long(df, ['A(quarterly)', 'B(quarterly)'],
+                        i='id', j='year')
+             X     A(quarterly)  B(quarterly)
+    id year
+    0  2010  0       0.531828       0.322959
+    1  2010  2       0.634401       0.361789
+    2  2010  2       0.849432       0.228263
+    0  2011  0       0.724455       0.293714
+    1  2011  2       0.611024       0.630976
+    2  2011  2       0.722443       0.092105
+
+    If we have many columns, we could also use a regex to find our
+    stubnames and pass that list on to wide_to_long
+
+    >>> stubnames = set([match[0] for match in
+                        df.columns.str.findall('[A-B]\(.*\)').values
+                        if match != [] ])
+    >>> list(stubnames)
+    ['B(quarterly)', 'A(quarterly)']
+
     Notes
     -----
     All extra variables are treated as extra id variables. This simply uses
     `pandas.melt` under the hood, but is hard-coded to "do the right thing"
     in a typicaly case.
     """
+    # For robustess, escape every user input string we use in a regex
+    import re
+
+    # We assume every 'varying' column in the data frame is of the format
+    # STUBsepPOSTFIX where STUB is the long format format varible name,
+    # sep is an optional single character delimiter and POSTFIX is the varying
+    # identifier. From this we assume we can identify the remaining
+    # non-varying variable names (id_vars)
 
     def get_var_names(df, regex):
         return df.filter(regex=regex).columns.tolist()
 
-    def melt_stub(df, stub, i, j):
-        varnames = get_var_names(df, "^" + stub)
-        newdf = melt(df, id_vars=i, value_vars=varnames, value_name=stub,
-                     var_name=j)
-        newdf_j = newdf[j].str.replace(stub, "")
-        try:
-            newdf_j = newdf_j.astype(int)
-        except ValueError:
-            pass
-        newdf[j] = newdf_j
-        return newdf
-
-    id_vars = get_var_names(df, "^(?!%s)" % "|".join(stubnames))
+    def melt_stub(df, stub, i, j, value_vars, sep):
+        newdf = melt(df, id_vars=i, value_vars=value_vars,
+                     value_name=stub.rstrip(sep), var_name=j)
+        newdf[j] = Categorical(newdf[j])
+        newdf[j] = newdf[j].str.replace(re.escape(stub), "")
+        return newdf.set_index([i, j])
+
+    stubs = list(map(lambda x: x + sep, stubnames))
+    # This regex is needed to avoid multiple "greedy" matches with stubs
+    # that have overlapping substrings
+    # (for example A2011, A2012 are separate from AA2011, AA2012)
+    value_vars = list(map(lambda x: get_var_names(
```

Comment:
But in the case of string stems, the three groups here will not be captured:

`Aone, Atwo, Bone, Btwo, BBone, BBtwo`

A negative lookahead `^B(?!B)` could be more robust, but not perfect...